### ISSUE TITLE
L2 cache throw exception when query with custom ManyToMany relation model

### ISSUE THREAD

**Issue by nguyenmcan** [ebean-orm/ebean#1722]
+ Java 8
+ Using Hazelcast for L2 cache 

### Steps to reproduce

```java
@Cache
@Entity
public class Address {
  @Id
  private long id;

  private String name;
}

@Cache
@Entity
public class Customer {
  @Id
  private long id;

  private String name;

  @OrderColumn(name = "priority")
  @OneToMany(cascade = CascadeType.ALL)
  private List<CustomerAddress> customerAddresses;
}

@Embeddable
public class CustomerAddressId {
  private long customerId;
  private long addressId;
}

@IdClass(CustomerAddressId.class)
@Entity
public class CustomerAddress {
  @Id
  @JoinColumn(name = "address_id")
  @ManyToOne
  private Address address;

  @Id
  @JoinColumn(name = "customer_id")
  @ManyToOne
  private Customer customer;

  public CustomerAddress() {
  }

  public CustomerAddress(Address address, Customer customer) {
    this.address = address;
    this.customer = customer;
  }
}

/**
    NOTE: The exception issue with the second query and `customerAddresses` must be empty

**/
  @Test
  public void findById() {
    Customer customer1 = Ebean.find(Customer.class).setUseCache(true).where().eq("id", 1).findOne();

    customer1.getCustomerAddresses().forEach(customerAddress -> {
      System.out.println(customerAddress.getAddress());
    });

    Customer customer2 = Ebean.find(Customer.class).setUseCache(true).where().eq("id", 1).findOne();

    customer2.getCustomerAddresses().forEach(customerAddress -> {
      System.out.println(customerAddress.getAddress());
    });
  }

```

```console
javax.persistence.PersistenceException: Error readSet on org.example.domain.CustomerAddressId.customerId
	at io.ebeaninternal.server.deploy.BeanProperty.readSet(BeanProperty.java:674)
	at io.ebeaninternal.server.deploy.id.IdBinderEmbedded.read(IdBinderEmbedded.java:352)
	at io.ebeaninternal.server.deploy.id.IdBinderEmbedded.readSet(IdBinderEmbedded.java:368)
	at io.ebeaninternal.server.query.SqlTreeNodeBean.load(SqlTreeNodeBean.java:265)
	at io.ebeaninternal.server.query.SqlTreeNodeManyRoot.load(SqlTreeNodeManyRoot.java:36)
	at io.ebeaninternal.server.query.SqlTreeNodeBean.load(SqlTreeNodeBean.java:328)
	at io.ebeaninternal.server.query.CQuery.readNextBean(CQuery.java:462)
	at io.ebeaninternal.server.query.CQuery.hasNext(CQuery.java:544)
	at io.ebeaninternal.server.query.CQuery.readBean(CQuery.java:519)
	at io.ebeaninternal.server.query.CQueryEngine.find(CQueryEngine.java:443)
	at io.ebeaninternal.server.query.DefaultOrmQueryEngine.findId(DefaultOrmQueryEngine.java:173)
	at io.ebeaninternal.server.core.OrmQueryRequest.findId(OrmQueryRequest.java:408)
	at io.ebeaninternal.server.core.DefaultServer.findId(DefaultServer.java:1233)
	at io.ebeaninternal.server.core.DefaultServer.findOne(DefaultServer.java:1251)
	at io.ebeaninternal.server.core.DefaultBeanLoader.loadManyInternal(DefaultBeanLoader.java:172)
	at io.ebeaninternal.server.core.DefaultBeanLoader.loadMany(DefaultBeanLoader.java:97)
	at io.ebeaninternal.server.core.DefaultServer.loadMany(DefaultServer.java:579)
	at io.ebean.common.AbstractBeanCollection.lazyLoadCollection(AbstractBeanCollection.java:98)
	at io.ebean.common.BeanList.init(BeanList.java:135)
	at io.ebean.common.BeanList.iterator(BeanList.java:329)
	at java.lang.Iterable.forEach(Iterable.java:74)
	at org.example.domain.CustomerQueryTest.findById(CustomerQueryTest.java:49)
```

**Comment by rbygrave** [comment]
Can you please create a project with this failing test that can be downloaded or git cloned.

Thanks, Rob.

**Comment by nguyenmcan** [comment]
Please checkout this and run test

https://github.com/nguyenmcan/ebean-orm-example.git

Thanks

**Comment by rbygrave** [comment]
Great thanks.  I should get to it tomorrow night.

**Comment by rbygrave** [comment]
Ok, this is next on the list now.

**Comment by rbygrave** [comment]
Right so why the NPE ... the customerId and addressId are primitive long values so their internal setter methods are taking primitive long.  The DB resultSet contains nulls so the null Long is implicitly converted to primitive long ... and we get the NPE.

```java
@Embeddable
public class ProductCategoryId {
  private long customerId;
  private long addressId;
```
So if they are object Long then the test passes:
```java
@Embeddable
public class ProductCategoryId {
  private Long customerId;
  private Long addressId;
```
Looking more ...

**Comment by rbygrave** [comment]
So pushed fix.  Handle the case for EmbeddedId of primitive values loaded from null value.  The fix changes from using `prop.readSet()` to `prop.read()` and checking for null.

**Comment by rbygrave** [comment]
Note that I have updated and released ebean-hazelcast to 11.39.1 (so you can update to the latest to get this fix if desired or change to `Long` types in the meantime.

**Comment by nguyenmcan** [comment]
Yup, 
Thanks @rbygrave Rob very much!
